### PR TITLE
4.next cake5

### PR DESCRIPTION
**Pull request by arusinowski**
Hi, this is `lorenzo/audit-stash` plugin updated to CakePHP 5.0, please review. To merge this PR into this repository the `4.next-cake5` or similar will need to be created. Thanks

### CODE REVIEW

**Comment by swiffer** [comment]
Can this be updated and released as 5.0 ist stable now?

**Comment by lorenzo** [comment]
There are conflicts in this branch

**Comment by aa-matthias** [comment]
@arusinowski do you continue working on the PR?

**Comment by arusinowski** [comment]
@aa-matthias  yes, i may continue on this but i will start again to resolve the conflicts properly, it may take couple of days.

**Comment by swiffer** [comment]
hi @arusinowski - sorry pinging you again, do you still plan to rework the PR?

**Comment by arusinowski** [comment]
> hi @arusinowski - sorry pinging you again, do you still plan to rework the PR?

Yes, tomorrow i will push here the code for a review (UTC evening).

**Comment by arusinowski** [comment]
Please review this PR, thank you

**Comment by swiffer** [comment]
thanks!

tried the new version. for me it's not logging a change, however no errors are thrown as well. will diagnose later today.

**Comment by cnizzardini** [comment]
Did you try running the unit tests? It also looks like @lorenzo needs approve the workflow.

**Comment by arusinowski** [comment]
@swiffer after the upgrade i tried with db and with elasticsearch as a backend and in both i was able to see the changes, what is your setup?

File: README.md [hunk 1]
```diff
@@ -274,7 +284,9 @@ The configuration contains the fully namespaced class name of your persister.
 
 ### Working With Transactional Queries
 
-Occasionally, you may want to wrap a number of database changes in a transaction, so that it can be rolled back if one part of the process fails. In order to create audit logs during a transaction, some additional setup is required. First create the file `src/Model/Audit/AuditTrail.php` with the following:
+Occasionally, you may want to wrap a number of database changes in a transaction, so that it can be rolled back if one
+part of the process fails. In order to create audit logs during a transaction, some additional setup is required. First
+create the file `src/Model/Audit/AuditTrail.php` with the following:
 
 ```php
 <?php
```
Comment (swiffer) [inline README.md:278]: declare(strict_types=1);

File: README.md [hunk 2]
```diff
@@ -333,4 +343,5 @@ if ($success) {
 }
```
Comment (swiffer) [inline README.md:336]: afterCommit expects param 3 to be of type ArrayObject

File: README.md [hunk 3]
```diff
@@ -315,16 +327,14 @@ Your transaction should then look similar to this example of a BookmarksControll
 
 ```php
 $trail = new AuditTrail();
-$success = $this->Bookmarks->connection()->transactional(function () use ($trail) {
+$success = $this->Bookmarks->connection()->transactional(function () use ($trail): void {
```
Comment (swiffer) [inline README.md]: i think something can be returned - why this change ?

File: README.md (hunk 2 above)
Comment (swiffer) [reply README.md:336]: $table->behaviors() should be $this->Bookmarks to match the previous line, or not?

File: src/Persister/RabbitMQPersister.php
Comment (cnizzardini) [inline src/Persister/RabbitMQPersister.php:1]: Why is this being deleted? Do people use this in the existing version?

File: src/Persister/DatabasePersister.php
Comment (cnizzardini) [inline src/Persister/DatabasePersister.php:1]: There is no need for this now or why is this being removed?

**Comment by swiffer** [comment]
> @swiffer after the upgrade i tried with db and with elasticsearch as a backend and in both i was able to see the changes, what is your setup?

i'm using the TablePersister - I think the case I tested never worked before as I was missing the transactional part of the README - so looks good for me now. Added some comments in line to improve the docs

**Comment by cnizzardini** [comment]
This should be updated as part of this PR: https://github.com/lorenzo/audit-stash/blob/master/.github/workflows/pull-request.yml

File: src/Persister/DatabasePersister.php
Comment (arusinowski) [reply src/Persister/DatabasePersister.php:1]: It was deprecated in previous version in favor of `\AuditStash\Persister\TablePersister` , this can be added to the release notes so users can adjust their code

File: README.md (hunk 3 above)
Comment (arusinowski) [reply README.md]: According to cakephp docs the closure do not need to return anything to make it work, but for clarity i can remove `void` as return type from this example

File: src/Persister/RabbitMQPersister.php
Comment (arusinowski) [reply src/Persister/RabbitMQPersister.php:1]: This was pushing the changes to rabbitmq but nothing was happening after that, i am unsure if that worked, this can be replaced with CakePHP queue.

Comment (cnizzardini) [reply src/Persister/RabbitMQPersister.php:1]: I guess this should be called out in the release notes too.

**Comment by cnizzardini** [comment]
We also need to create 3.x branch for cakephp 4.x users.

**Comment by swiffer** [comment]
Hey - Checking in again - can that be merged & released?

**Comment by jharder** [comment]
Hi there! I found a fork of this (kkdesilva/audit-trail) that added some nice functionality and while I was using it, I saw that you've completed the CakePHP5 upgrade!

I rebased off of this PR and just pushed my changes. I still have to write a few tests to cover the new functionality but otherwise everything seems to be working (I even set up ElasticSearch to test that part!)

Should I wait until this PR is accepted, or can we merge them both and only take one pull?

**Comment by swiffer** [comment]
@lorenzo  / @cnizzardini - could we get this released to ensure the great work done by @arusinowski is integrated?
As @jharder is looking to extend functionality afterwards it looks like there is a certain momentum in the community that shouldn't be missed.

**Comment by arusinowski** [comment]
> Hi there! I found a fork of this (kkdesilva/audit-trail) that added some nice functionality and while I was using it, I saw that you've completed the CakePHP5 upgrade!
> 
> I rebased off of this PR and just pushed my changes. I still have to write a few tests to cover the new functionality but otherwise everything seems to be working (I even set up ElasticSearch to test that part!)
> 
> Should I wait until this PR is accepted, or can we merge them both and only take one pull?

That's great, i think we can wait until this PR is accepted. I will look at your fork in the next week. Thanks

**Comment by cnizzardini** [comment]
> @lorenzo / @cnizzardini - could we get this released to ensure the great work done by @arusinowski is integrated? As @jharder is looking to extend functionality afterwards it looks like there is a certain momentum in the community that shouldn't be missed.

I am not a maintainer. @lorenzo needs to make me one and ideally at least one other person so we can start getting these branches merged in.

**Comment by dereuromark** [comment]
maybe @lorenzo is open to having experienced maintainers helping him with the maintenance of the plugin.
It seems there is quite some interest in having this in Cake5 as well as maintained in general.

**Comment by cnizzardini** [comment]
@dereuromark someone needs to do a wellness check on @lorenzo as he is MIA.

**Comment by swiffer** [comment]
any news ?

**Comment by dereuromark** [comment]
Let me cast a resurrection spell : Hokus Pokus

**Comment by lorenzo** [comment]
@dereuromark it worked!!

**Comment by lorenzo** [comment]
@arusinowski thanks for your work! Let me know if you would like to have maintainer rights in this repo

**Comment by arusinowski** [comment]
@lorenzo Thank you for accepting my PR :) . As for being a maintainer here, yes, I would like to be one.

**Comment by lorenzo** [comment]
@arusinowski invitation sent

**Comment by cnizzardini** [comment]
This branch broke being able to run tests locally and my guess was not really validated much sigh, gonna need to put better quality controls on this project. Running a full blown elastic instance on my local shouldn't be a requirement for developing on this...


```
PHPUnit 10.5.5 by Sebastian Bergmann and contributors

Error in bootstrap script: Elastica\Exception\Connection\HttpException:
Couldn't resolve host
#0 /home/chris/dev/c-audit-stash/vendor/ruflin/elastica/src/Request.php(183): Elastica\Transport\Http->exec()
#1 /home/chris/dev/c-audit-stash/vendor/ruflin/elastica/src/Client.php(545): Elastica\Request->send()
#2 /home/chris/dev/c-audit-stash/vendor/ruflin/elastica/src/Client.php(580): Elastica\Client->request()
#3 /home/chris/dev/c-audit-stash/vendor/ruflin/elastica/src/Index.php(774): Elastica\Client->requestEndpoint()
#4 /home/chris/dev/c-audit-stash/vendor/ruflin/elastica/src/Index.php(522): Elastica\Index->requestEndpoint()
#5 /home/chris/dev/c-audit-stash/vendor/cakephp/elastic-search/src/TestSuite/Fixture/MappingGenerator.php(92): Elastica\Index->exists()
#6 /home/chris/dev/c-audit-stash/vendor/cakephp/elastic-search/src/TestSuite/Fixture/MappingGenerator.php(77): Cake\ElasticSearch\TestSuite\Fixture\MappingGenerator->dropIndex()
#7 /home/chris/dev/c-audit-stash/tests/bootstrap.php(55): Cake\ElasticSearch\TestSuite\Fixture\MappingGenerator->reload()
#8 /home/chris/dev/c-audit-stash/vendor/phpunit/phpunit/src/TextUI/Application.php(283): include_once('...')
#9 /home/chris/dev/c-audit-stash/vendor/phpunit/phpunit/src/TextUI/Application.php(103): PHPUnit\TextUI\Application->loadBootstrapScript()
#10 /home/chris/dev/c-audit-stash/vendor/phpunit/phpunit/phpunit(99): PHPUnit\TextUI\Application->run()
#11 /home/chris/dev/c-audit-stash/vendor/bin/phpunit(122): include('...')
#12 {main}

```

**Comment by cnizzardini** [comment]
Further, it doesn't even seem to pass the pipeline... How did this get merged?
